extends Node2D
@onready var ç™»å½•ç”¨æˆ·å: LineEdit = $"Control/ç™»å½•ç•Œé¢/VBoxContainer/MarginContainer/äº’åŠ¨å±‚/ç”¨æˆ·å/ç”¨æˆ·å/ç”¨æˆ·è¾“å…¥æ¡†"
@onready var ç™»å½•å¯†ç : LineEdit = $"Control/ç™»å½•ç•Œé¢/VBoxContainer/MarginContainer/äº’åŠ¨å±‚/å¯†ç /å¯†ç /å¯†ç è¾“å…¥æ¡†"
@onready var æ³¨å†Œç”¨æˆ·å: LineEdit = $"Control/æ³¨å†Œç•Œé¢/VBoxContainer/MarginContainer/äº’åŠ¨å±‚/ç”¨æˆ·å/ç”¨æˆ·å/ç”¨æˆ·è¾“å…¥æ¡†"
@onready var æ³¨å†Œå¯†ç : LineEdit = $"Control/æ³¨å†Œç•Œé¢/VBoxContainer/MarginContainer/äº’åŠ¨å±‚/å¯†ç /å¯†ç /å¯†ç è¾“å…¥æ¡†"

@onready var æç¤ºåœºæ™¯: PanelContainer = $"Control/æ³¨å†Œç•Œé¢/VBoxContainer/æç¤º/æç¤ºåœºæ™¯"
@onready var ç™»å½•ç•Œé¢: MarginContainer = $"Control/ç™»å½•ç•Œé¢"
@onready var æ³¨å†Œç•Œé¢: MarginContainer = $"Control/æ³¨å†Œç•Œé¢"
@onready var æ³¨å†Œ: Button = $"Control/ç™»å½•ç•Œé¢/VBoxContainer/MarginContainer/äº’åŠ¨å±‚/æ“ä½œå±‚/æŒ‰é’®/æ³¨å†ŒæŒ‰é’®"
@onready var ç™»å½•: Button = $"Control/ç™»å½•ç•Œé¢/VBoxContainer/MarginContainer/äº’åŠ¨å±‚/æ“ä½œå±‚/æŒ‰é’®/ç™»å½•æŒ‰é’®"
@onready var kapibala: AnimatedSprite2D = $"å¡çš®å·´æ‹‰"

var remind = {
	1: "æç¤º1ï¼šå¯†ç éœ€è¦åŒ…å«æ•°å­—",
	2: "æç¤º2ï¼šå¯†ç éœ€è¦åŒ…å«å¤§å†™å­—æ¯",
	3: "æç¤º3ï¼šå¯†ç éœ€è¦åŒ…å«è‡³å°‘ä¸€ä¸ªç‰¹æ®Šç¬¦å·",
	4: "æç¤º4ï¼šå¯†ç ä¸­æ•°å­—å’Œæ•°å­—ä¹‹é—´è‹¥å·®å°äº4ï¼Œå¿…é¡»ä¸­é—´éš”ä¸€ä¸ªå­—æ¯",
	5: "æç¤º5ï¼šæ‚¨çš„å¯†ç ä¸è¶³9ä½",
	6: "æç¤º6ï¼šæ‚¨çš„å¯†ç ä¸å¤Ÿå¼ºå£®"
}
var animation = {
	1: "grass",
	2: "walk",
	3: "sitting_1",
	4: "sitting_2"
}
var account:Dictionary  = {}
var flag:bool = false #æ˜¯å¦å¯ä»¥æ³¨å†Œ


func _ready() -> void:
	æç¤ºåœºæ™¯.set_text(remind.get(1))
	kapibala_play()
	
func kapibala_play():
	kapibala.play(animation[randi_range(1,animation.size())])
	
func æ£€éªŒ(s:String)->int:
	if !æç¤º1(s):
		return 1
	if !æç¤º2(s):
		return 2
	if !æç¤º3(s):
		return 3
	if !æç¤º4(s):
		return 4
	if !æç¤º5(s):
		return 5	
	if !æç¤º6(s):
		return 6	
	return 0
	
func æç¤º1(s:String)->bool:
	return æ£€æµ‹æ˜¯å¦æœ‰æ•°å­—(s)

func æç¤º2(s:String)->bool:
	return æ£€æµ‹æ˜¯å¦æœ‰å¤§å†™å­—æ¯(s)
	
func æç¤º3(s:String)->bool:
	var regex = RegEx.new()
	regex.compile("[!@#$%^&(),.?\":{}|<>]")
	return regex.search(s)!=null
	
func æç¤º4(s:String)->bool:
	var regex = RegEx.new()
	regex.compile("([0-9])([0-9])")
	var result = regex.search(s)
	if result:
		var first_number = int(result.get_strings()[1])
		var second_number = int(result.get_strings()[2])
		if abs(first_number - second_number) == 4:
			return true
		else:
			return false
	return true;
	
func æç¤º5(s:String)->bool:
	return s.length() >= 9;
	
func æç¤º6(s:String)->bool:
	var emoji_count = 0
	for char in s:
		if char == "ğŸ’ª":
			emoji_count += 1
		if emoji_count >= 3:
			return true
	return false
	
func æ£€æµ‹æ˜¯å¦æœ‰æ•°å­—(s:String)->bool:
	var regex = RegEx.new()
	regex.compile("\\d")
	return regex.search(s)!=null
	
func æ£€æµ‹æ˜¯å¦æœ‰å¤§å†™å­—æ¯(s:String)->bool:
	var regex = RegEx.new()
	regex.compile("[A-Z]")
	return regex.search(s)!=null

func æ£€æµ‹æ˜¯å¦æœ‰å°å†™å­—æ¯(s:String)->bool:
	var regex = RegEx.new()
	regex.compile("[a-z]")
	return regex.search(s)!=null
		
#æ³¨å†Œç¡®è®¤
func _on_ç¡®è®¤æŒ‰é’®_pressed() -> void:
	if flag:
		account.get_or_add(æ³¨å†Œç”¨æˆ·å.text,æ³¨å†Œå¯†ç .text)
		æ³¨å†Œç•Œé¢.visible = false
		ç™»å½•ç•Œé¢.visible = true
		kapibala_play()
		ç™»å½•ç”¨æˆ·å.text = ""
		ç™»å½•å¯†ç .text = ""

func _on_æ³¨å†ŒæŒ‰é’®_pressed() -> void:
	æ³¨å†Œç•Œé¢.visible = true
	ç™»å½•ç•Œé¢.visible = false
	kapibala_play()
	æ³¨å†Œå¯†ç .text = "ğŸ’ª"
	æ³¨å†Œç”¨æˆ·å.text = ""
func _on_ç™»å½•æŒ‰é’®_pressed() -> void:
	GlobalGameManager.emit_complete_game()

func _on_è¿”å›æŒ‰é’®_pressed() -> void:
	æ³¨å†Œç•Œé¢.visible = false
	ç™»å½•ç•Œé¢.visible = true
	kapibala_play()
	ç™»å½•ç”¨æˆ·å.text = ""
	ç™»å½•å¯†ç .text = ""

func _on_å¯†ç è¾“å…¥æ¡†_text_changed(new_text: String) -> void:
	#æç¤ºæ–‡å­—çš„keyå€¼
	var text_index = æ£€éªŒ(new_text)
	if remind.get(text_index):
		flag = false
		æç¤ºåœºæ™¯.set_text(remind.get(text_index))
	else:
		flag = true	
		æç¤ºåœºæ™¯.set_text("")
